firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -345,28 +359,37 @@ private Void registerAndSaveFid(PersistedInstallationEntry persistedInstallation
   }
 
   /** Calls the FIS servers to generate an auth token for this Firebase installation. */
-  private InstallationTokenResult fetchAuthTokenFromServer(
-      PersistedInstallationEntry persistedInstallationEntry) throws FirebaseInstallationsException {
+  private void fetchAuthTokenFromServer(PersistedInstallationEntry persistedInstallationEntry)
+      throws FirebaseInstallationsException {
     try {
       long creationTime = utils.currentTimeInSecs();
-      InstallationTokenResult tokenResult =
+      TokenResult tokenResult =
           serviceClient.generateAuthToken(
               /*apiKey= */ firebaseApp.getOptions().getApiKey(),
               /*fid= */ persistedInstallationEntry.getFirebaseInstallationId(),
               /*projectID= */ firebaseApp.getOptions().getProjectId(),
               /*refreshToken= */ persistedInstallationEntry.getRefreshToken());
 
-      persistedInstallation.insertOrUpdatePersistedInstallationEntry(
-          PersistedInstallationEntry.builder()
-              .setFirebaseInstallationId(persistedInstallationEntry.getFirebaseInstallationId())
-              .setRegistrationStatus(RegistrationStatus.REGISTERED)
-              .setAuthToken(tokenResult.getToken())
-              .setRefreshToken(persistedInstallationEntry.getRefreshToken())
-              .setExpiresInSecs(tokenResult.getTokenExpirationTimestamp())
-              .setTokenCreationEpochInSecs(creationTime)
-              .build());
-
-      return tokenResult;
+      if (tokenResult.getResponseCode() == TokenResult.ResponseCode.FID_ERROR
+          || tokenResult.getResponseCode() == TokenResult.ResponseCode.REFRESH_TOKEN_ERROR) {
+        persistedInstallation.insertOrUpdatePersistedInstallationEntry(
+            persistedInstallationEntry
+                .toBuilder()
+                .setRegistrationStatus(RegistrationStatus.AUTHENTICATION_TOKEN_ERROR)
+                .build());
```

Comment:
In the auth error case, if anyways you will clear the cache, why do you update it to AUTHENTICATION_TOKEN_ERROR and then clear it in doRegistration()?

An alternative is to clear it here, and like your old code, still return TokenResult from this method.

Reply:
Makes sense. Changed it. However, I retained the return type as void because it is unused by the caller of the method.